File: C++/Recursed/collatz.cpp
// Hannah Gulle
// Computes the top 10 starting values from 2 to a given positive index
// by collatz sequence length and starting value recursively

#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

// collatz struct holds sequence start and sequence length
struct Collatz{
	unsigned long long int start;
	unsigned long long int length;
};

// Sorting Methods
bool startSort(Collatz const& lhs, Collatz const& rhs);
bool lengthSort(Collatz const& lhs, Collatz const& rhs);

// Collatz Computation Method
vector<unsigned long long int> collatz( vector<unsigned long long int> sequence, unsigned long long int start );

// Main Program
int main(){

	// Retrieves the maximum starting integer from the keyboard
	unsigned long long int max;
	cout << "Enter the Highest Starting Integer: ";
	cin >> max;

	// vector of all collatz objects within computation range
	vector<Collatz> all;
	// vector of all collatz sequence values for a given starting value
	vector<unsigned long long int> sequence;
	// length of a given sequence
	unsigned long long int length;
	
	// computes the collatz sequence for each starting value within range
	for(unsigned long long int i = 2; i < max; i++ ){
		sequence.push_back(i);
		sequence = collatz( sequence, i );
	
		length = sequence.size()-1;
		Collatz c = {i, length};
		all.push_back(c);

		sequence.clear();
	}

	cout << "Top 10 Starting Values After Sorting by Length" << endl;
	// sorts all collatz objects by sequence length
	sort( all.begin(), all.end(), &lengthSort );

	// Calculates and Outputs the top 10 starting values after sorting by length
	int top = 1;
	vector<Collatz> topTen;
	for(int i = all.size()-1; i > 1; i--){
		if( all[i].length > all[i-1].length){
			topTen.push_back( all[i] );
			cout << all[i].start << " " << all[i].length << endl;
			top++;
		}
		
		if( top > 10 ){
			break;
		}
	}	

	cout << "\n Top 10 Starting Values After Sorting by Start" << endl;
	// sorts top 10 collatz objects by sequence start
	sort( topTen.begin(), topTen.end(), &startSort );

	for( int i = 9; i > 0; i-- ){
		cout << topTen[i].start << " " << topTen[i].length << endl;
	}
	
	return 0;
}

// Sorting Method: Sorts collatz objects by starting value
bool startSort(Collatz const& lhs, Collatz const& rhs){
	return lhs.start < rhs.start;
}

// Sorting Method: sorts collatz objects by length
bool lengthSort(Collatz const& lhs, Collatz const& rhs){
	return lhs.length < rhs.length;
}

// recursively determines the collatz sequence for a given starting value
vector<unsigned long long int> collatz( vector<unsigned long long int> sequence, unsigned long long int start ){

	if( start == 1 ){
		return sequence;
	}
	else if( start % 2 == 0 ){
		sequence.push_back( start/2 );
		return collatz( sequence, start/2 );
	}
	else{
		sequence.push_back( (3 * start) + 1 );
		return collatz( sequence, (3 * start) + 1 );
	}
}

File: Julia/Recursed/collatz.jl
# Hannah gulle
# Program computes the top 10 collatz seqeunces for 
# a given starting range

# collatz type holds sequence start and sequence size
type Collatz
	start	:: Int
	size	:: Int
end;

# computes the collatz seqeucne size for a given
# starting integer recursively
function getCollatz(seq, start::Int)

	if start == 1 
		return seq
	elseif start%2 == 0
		push!(seq, Int(start/2))
		return getCollatz( seq, Int(start/2))
	else
		push!(seq, Int((3*start)+1))
		return getCollatz( seq, Int((3*start)+1))
	end
end

# returns the size of a given collatz object
function getSize( c::Collatz)
	return c.size
end

# returns the start of a given collatz object
function getStart( c::Collatz )
	return c.start
end

# request the range upper bound from the user
println("Input the Highest Starting Integer")
endVal = readline()
endVal = parse(Int, endVal)

# create an array of collatz types to hold all
# collatz sequence size and start values
allseq = Collatz[]

# iterate through the range to find the collatz
# seqeunce size
for i= 2:endVal

	sequenz = [i]
	getCollatz(sequenz, i)
	size = countnz(sequenz)-1

	c = Collatz(i, size)
	push!(allseq, c )
	empty!(sequenz)
end

println("")
println("Top 10 Starting Values After Sorting by Length")

# sort the array of collatz types by seqeunce size
sort!( allseq, by=getSize)

# array to hold top 10 values
top = 1
topTen = Collatz[]

# output top 10 values sorted by size
for i= countnz(allseq):-1:1
	if allseq[i].size > allseq[i-1].size
		push!( topTen, allseq[i] )
		println( allseq[i].start, " ", allseq[i].size)
		top = top+1
	end
	if top > 10
		break
	end
end

println("")
println("Top 10 Starting Values After Sorting by Start")

# sort the top 10 values by start
sort!( topTen, by=getStart)

# output top 10 after sorting by start
for i=10:-1:1
	println( topTen[i].start, " ", topTen[i].size)
end
